File: Kotlin/src/main/kotlin/com/gildedrose/GildedRose.kt
package com.gildedrose

class GildedRose(var items: Array<Item>) {
    fun updateQuality() {
        items.forEach { oldItem ->
            val item = when {
                oldItem.name == "Sulfuras, Hand of Ragnaros" -> LegendaryItem(oldItem)
                oldItem.name == "Aged Brie" -> AgedBrie(oldItem)
                oldItem.name == "Backstage passes to a TAFKAL80ETC concert" -> BackStagePass(oldItem)
                oldItem.name.contains("Conjured") -> ConjuredItem(oldItem)
                else -> BasicItem(oldItem)
            }

            item.updateQuality()
            item.age()

            if (item.isExpired()) {
                item.updateQuality()
            }

            item.copyInto(oldItem)
        }
    }

    open class BasicItem(item: Item) : Item(item.name, item.sellIn, item.quality) {
        fun copyInto(item: Item) {
            item.sellIn = sellIn
            item.quality = quality
            item.name = name
        }

        private fun belowMaximumQuality() = quality < 50
        private fun aboveMinimumQuality() = quality > 0

        fun increaseQuality() {
            if (belowMaximumQuality()) {
                quality += 1
            }
        }

        fun decreaseQuality() {
            if (aboveMinimumQuality()) {
                quality -= 1
            }
        }

        open fun updateQuality() = decreaseQuality()

        open fun age() {
            sellIn -= 1
        }

        open fun isExpired() = sellIn < 0
    }

    class LegendaryItem(item: Item) : BasicItem(item) {
        override fun updateQuality() = Unit
        override fun age() = Unit
    }

    class AgedBrie(item: Item) : BasicItem(item) {
        override fun updateQuality() {
            super.increaseQuality()
        }
    }

    class BackStagePass(item: Item) : BasicItem(item) {
        private fun expiresIn(days: Int) = sellIn <= days

        override fun updateQuality() {
            if (isExpired()) {
                quality = 0
                return
            }

            super.increaseQuality()
            if (expiresIn(10)) {
                super.increaseQuality()
            }

            if (expiresIn(5)) {
                super.increaseQuality()
            }
        }
    }

    class ConjuredItem(item: Item) : BasicItem(item) {
        override fun updateQuality() {
            super.decreaseQuality()
            super.decreaseQuality()
        }
    }
}

File: TypeScript/app/gilded-rose.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

class BasicItem extends Item {
  constructor(item: Item) {
    super(item.name, item.sellIn, item.quality);
  }

  copyInto(item: Item) {
    item.name = this.name;
    item.sellIn = this.sellIn;
    item.quality = this.quality;
  }

  isAgedBrie(): boolean {
    return this.name === "Aged Brie";
  }

  isLegendaryItem(): boolean {
    return this.name === "Sulfuras, Hand of Ragnaros";
  }

  isBackstagePass(): boolean {
    return this.name === "Backstage passes to a TAFKAL80ETC concert";
  }

  age(): void {
    this.sellIn--;
  }

  increaseQuality(): void {
    this.quality++;
  }

  decreaseQuality(): void {
    this.quality--;
  }
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  updateQuality() {
    this.items.forEach((oldItem: Item) => {
      const item = new BasicItem(oldItem);

      if (!item.isAgedBrie() && !item.isBackstagePass()) {
        if (item.quality > 0) {
          if (!item.isLegendaryItem()) {
            item.decreaseQuality();
          }
        }
      } else {
        if (item.quality < 50) {
          item.increaseQuality();
          if (item.isBackstagePass()) {
            if (item.sellIn < 11) {
              if (item.quality < 50) {
                item.increaseQuality();
              }
            }
            if (item.sellIn < 6) {
              if (item.quality < 50) {
                item.increaseQuality();
              }
            }
          }
        }
      }

      if (!item.isLegendaryItem()) {
        item.age();
      }

      if (item.sellIn < 0) {
        if (!item.isAgedBrie()) {
          if (!item.isBackstagePass()) {
            if (item.quality > 0) {
              if (!item.isLegendaryItem()) {
                item.decreaseQuality();
              }
            }
          } else {
            item.quality = 0;
          }
        } else {
          if (item.quality < 50) {
            item.increaseQuality();
          }
        }
      }

      item.copyInto(oldItem);
    });

    return this.items;
  }
}
